azure-iot-device/azure/iot/device/provisioning/internal/polling_machine.py

```diff
@@ -0,0 +1,325 @@
+                "source": "polling",
+                "dest": "error",
+                "after": "_call_error",
+            },
+            {"trigger": "_trig_cancel", "source": "disconnected", "dest": None},
+            {
+                "trigger": "_trig_cancel",
+                "source": "registering",
+                "dest": "cancelling",
+                "after": "clear_and_disconnect",
+            },
+            {
+                "trigger": "_trig_cancel",
+                "source": "waiting_to_poll",
+                "dest": "cancelling",
+                "after": "clear_and_disconnect",
+            },
+            {
+                "trigger": "_trig_cancel",
+                "source": "polling",
+                "dest": "cancelling",
+                "after": "clear_and_disconnect",
+            },
+        ]
+
+        def _on_transition_complete(event_data):
+            if not event_data.transition:
+                dest = "[no transition]"
+            else:
+                dest = event_data.transition.dest
+            logger.info(
+                "Transition complete.  Trigger=%s, Dest=%s, result=%s, error=%s",
+                event_data.event.name,
+                dest,
+                str(event_data.result),
+                str(event_data.error),
+            )
+
+        self._state_machine = Machine(
+            model=self,
+            states=states,
+            transitions=transitions,
+            initial="disconnected",
+            send_event=True,  # Use event_data structures to pass transition arguments
+            finalize_event=_on_transition_complete,
+            queued=True,
+        )
+
+    def register(self, callback=None):
+        """
+        Register the device with the provisioning service.
+        """
+        logger.info("register called from polling machine")
+        self._register_callback = callback
+        self._trig_register()
+
+    def cancel(self, callback=None):
+        logger.info("cancel called from polling machine")
+        self._cancel_callback = callback
+        self._trig_cancel()
+
+    def _initialize_register(self, event_data):
+        logger.info("Preparing for registration process.")
+        # Choice: Sending action here as the polling/dps state machine must figure out
+        # in what order action must be sent. The request response is only responsible for
+        # requesting and responding. However the actions can be taken down to the
+        # the request response machine as well.
+        subscribe_action = SubscribeAction(
+            constant.SUBSCRIBE_TOPIC_PROVISIONING, qos=1, callback=self.on_subscribe_completed
+        )
+        # The request response takes a tuple of actions
+        # It will call state based provider connect which
+        # It will add subscribe to pending action queue
+        # it will execute connect and then execute subscribe
+        # TODO : Should I have a waiting callback or a connack on which something happens ?
```

Comment:
># TODO : Should I have a waiting callback or a connack on which something happens ? [](start = 8, length = 83)

The state-based provider will automatically connect and subscribe when you issue a subscribe request.  Why not just call subscribe here and let the state-based machine do the connect?

Reply:
thats one way to go. 
Here I am connecting and the subscribe happens automatically as a post connect action.
But we can also subscribe and it will connect automatically.
It wont change however that only after SUBACK comes back I will send the request finally.

---
In reply to: [277398333](https://github.com/Azure/azure-iot-sdk-python-preview/pull/81#discussion_r277398333) [](ancestors = 277398333)